src/main/java/org/springframework/samples/petclinic/rest/importcsv/ImportCSV.java

```diff
@@ -26,114 +26,123 @@
+
     public ResponseEntity<List<Pet>> importPets(@RequestBody String csv) {
 
-        int i = 0;
         List<Pet> pets = new LinkedList<Pet>();
-        Pet pet;
 
-        do {
-            pet = new Pet();
+        try{
+            pets = ParseFields(csv);
+        } catch (IncorrectFormatException e) {
+            HttpHeaders headers = new HttpHeaders();
+            headers.add("errors", e.getMessage());
+            return new ResponseEntity<List<Pet>>(headers, HttpStatus.BAD_REQUEST);
+        }
 
-            String field = "";
-            while (i < csv.length() && csv.charAt(i) != ';') {
-                field += csv.charAt(i++);
-            }
-            i++;
-
-            pet.setName(field);
-
-            field = "";
-            while (i < csv.length() && csv.charAt(i) != ';') {
-                field += csv.charAt(i++);
-            }
-            i++;
-
-            try {
-                pet.setBirthDate((new SimpleDateFormat("yyyy-MM-dd")).parse(field));
-            } catch (ParseException e) {
-                HttpHeaders headers = new HttpHeaders();
-                headers.add("errors", "date " + field + " not valid");
-                return new ResponseEntity<List<Pet>>(headers, HttpStatus.BAD_REQUEST);
-            }
+        return new ResponseEntity<List<Pet>>(pets, HttpStatus.OK);
+    }
 
-            field = "";
-            while (i < csv.length() && csv.charAt(i) != ';') {
-                field += csv.charAt(i++);
-            }
-            i++;
-
-            if (pet != null) {
-                ArrayList<PetType> ts = (ArrayList<PetType>) clinicService.findPetTypes();
-                for (int j = 0; j < ts.size(); j++) {
-                    if (ts.get(j).getName().toLowerCase().equals(field)) {
-                        pet.setType(ts.get(j));
-                        break;
-                    }
-                }
-            }
+    /**
+     * Custom Exception which can arise during csv string parsing, error message
+     * contains where the exception was raised
+     */
+    private class IncorrectFormatException extends Exception {
+        public IncorrectFormatException(String errorMessage) {
+            super(errorMessage);
+        }
+    }
 
-            field = "";
-            while (i < csv.length() && (csv.charAt(i) != ';' && csv.charAt(i) != '\n')) {
-                field += csv.charAt(i++);
+    /**
+     * Method which receives a string from the csv and pars it to a list of pets
+     * which is then returns. Exceptions are thrown upon the error field
+     *
+     * @param csv the csv string from the user
+     * @return pets list of pets
+     * @throws IncorrectFormatException the user input is malformatted
+     */
+    private List<Pet> ParseFields(String csv) throws IncorrectFormatException {
+        int i = 0;
+        List<Pet> pets = new LinkedList<Pet>();
+        Pet pet;
+        pet = new Pet();
+
+        String[] arrayOfCsv = csv.split(";", 0);
```

Comment:
In my opinion, `arrayOfCsv` sounds like an array of CSV objects, where it is in fact an array of input fields. Maybe `fields` would be a better name. 

Also according to the API description, you could get multiple commands separated by `\n`. In this case you are only considering the first command and ignoring the rest. Can also be seen in line 122, where you only add one pet and return from the function.